charmcraft/commands/store/client.py

```diff
@@ -0,0 +1,144 @@
+# functionality in Store's production
+BASE_URL = 'https://api.staging.snapcraft.io/publisher/api'
+
+
+def visit_page_with_browser(visit_url):
+    """Open a browser so the user can validate its identity."""
+    logger.warning(
+        "Opening an authorization web page in your browser; if it does not open, "
+        "please open this URL: %s", visit_url)
+    webbrowser.open(visit_url, new=1)
+
+
+class _AuthHolder:
+    """Holder and wrapper of all authentication bits.
+
+    Main two purposes of this class:
+
+    - deal with credentials persistence
+
+    - wrap HTTP calls to ensure authentication
+
+    XXX Facundo 2020-06-18: right now for functionality bootstrapping we're storing credentials
+    on disk, we may move to a keyring, wallet, other solution, or firmly remain here when we
+    get a "security" recommendation.
+    """
+
+    def __init__(self):
+        self._cookiejar_filepath = appdirs.user_config_dir('charmcraft.credentials')
+        self._cookiejar = None
+        self._client = None
+
+    def clear_credentials(self):
+        """Clear stored credentials."""
+        if os.path.exists(self._cookiejar_filepath):
+            os.unlink(self._cookiejar_filepath)
+            logger.debug("Credentials cleared: file %r removed", self._cookiejar_filepath)
+        else:
+            logger.debug("Credentials file not found to be removed: %r", self._cookiejar_filepath)
+
+    def _save_credentials(self):
+        """Save credentials if changed."""
+        if list(self._cookiejar) != self._old_cookies:
+            logger.debug("Saving credentials to file: %r", self._cookiejar_filepath)
+            dirpath = os.path.dirname(self._cookiejar_filepath)
+            if not os.path.exists(dirpath):
+                os.makedirs(dirpath)
+
+            self._cookiejar.save()
+
+    def _load_credentials(self):
+        """Load credentials and set up internal auth request objects."""
+        wbi = httpbakery.WebBrowserInteractor(open=visit_page_with_browser)
+        self._cookiejar = MozillaCookieJar(self._cookiejar_filepath)
+        self._client = httpbakery.Client(cookies=self._cookiejar, interaction_methods=[wbi])
+
+        if os.path.exists(self._cookiejar_filepath):
+            logger.debug("Loading credentials from file: %r", self._cookiejar_filepath)
+            try:
+                self._cookiejar.load()
+            except Exception as err:
+                # alert and continue processing (without having credentials, of course, the user
+                # will be asked to authenticate)
+                logger.warning("Failed to read credentials: %r", err)
+        else:
+            logger.debug("Credentials file not found: %r", self._cookiejar_filepath)
+
+        # iterates the cookiejar (which is mutable, may change later) and get the cookies
+        # for comparison after hitting the endpoint
+        self._old_cookies = list(self._cookiejar)
```

Comment:
if it is mutable do we need a deep copy?

Reply:
Nop, the cookiejar is the mutable one, which may change having different cookies after the requests. Each internal cookie will not mutate.